src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.OpenSsl.cs

```diff
@@ -212,52 +203,46 @@ internal static int Encrypt(SafeSslHandle context, byte[] input, int offset, int
                     case Ssl.SslErrorCode.SSL_ERROR_ZERO_RETURN:
                     case Ssl.SslErrorCode.SSL_ERROR_WANT_READ:
                         break;
-
+                    // indicates we need to write the out buffer and write again
+                    case Ssl.SslErrorCode.SSL_ERROR_WANT_WRITE:
+                        break;
                     default:
                         throw new SslException(SR.Format(SR.net_ssl_encrypt_failed, errorCode), innerError);
                 }
             }
-            else
-            {
-                int capacityNeeded = Crypto.BioCtrlPending(context.OutputBio);
-
-                if (output == null || output.Length < capacityNeeded)
-                {
-                    output = new byte[capacityNeeded];
-                }
-
-                retVal = BioRead(context.OutputBio, output, capacityNeeded);
-            }
 
-            return retVal;
+            int bytesWritten = context.OutputBio.BytesWritten;
+            context.OutputBio.Reset();
+            return bytesWritten;
         }
 
         internal static int Decrypt(SafeSslHandle context, byte[] outBuffer, int offset, int count, out Ssl.SslErrorCode errorCode)
         {
+            Debug.Assert(offset >= 0);
+            Debug.Assert(offset <= outBuffer.Length);
+
             errorCode = Ssl.SslErrorCode.SSL_ERROR_NONE;
 
-            int retVal = BioWrite(context.InputBio, outBuffer, offset, count);
+            context.InputBio.SetData(outBuffer, offset, count);
 
-            if (retVal == count)
+            int retVal;
+            unsafe
             {
-                unsafe
+                fixed (byte* fixedBuffer = outBuffer)
                 {
-                    fixed (byte* fixedBuffer = outBuffer)
-                    {
-                        retVal = Ssl.SslRead(context, fixedBuffer + offset, outBuffer.Length);
-                    }
+                    retVal = Ssl.SslRead(context, fixedBuffer + offset, outBuffer.Length - offset);
```

Comment:
No this is our buffer we own internally, we can always use the entire thing. The count is the number of encrypted bytes currently in the same buffer for decrypting. This is the pattern for how it works pre the PR.